controller/Lists.php

```diff
@@ -524,40 +413,98 @@ public function removeList(?string $uri = null): void
         $this->returnJson(['deleted' => $deleted]);
     }
 
-    /**
-     * Remove the list element in parameter
-     * @throws common_exception_BadRequest
-     * @return void
-     */
-    public function removeListElement()
+    public function removeListElement(): void
     {
         if (!$this->isXmlHttpRequest()) {
             throw new common_exception_BadRequest('wrong request mode');
         }
+
         $deleted = false;
 
         if ($this->hasRequestParameter('uri')) {
             $deleted = $this->getListService()->removeListElement(
-                $this->getResource(tao_helpers_Uri::decode($this->getRequestParameter('uri')))
+                $this->getOntology()->getResource(
+                    tao_helpers_Uri::decode($this->getRequestParameter('uri'))
+                )
             );
         }
+
         $this->returnJson(['deleted' => $deleted]);
     }
 
-    /**
-     * @return ListService
-     */
-    protected function getListService()
+    private function createList(array $values): core_kernel_classes_Class
     {
-        return $this->getPsrContainer()->get(ListService::class);
+        $class = $this->getListService()->createList($values[tao_actions_form_RemoteList::FIELD_NAME]);
+
+        $propertyType = $class->getProperty(CollectionType::TYPE_PROPERTY);
+        $propertyRemote = $class->getProperty((string) CollectionType::remote());
+        $class->setPropertyValue($propertyType, $propertyRemote);
+
+        $propertySource = $class->getProperty(RemoteSourcedListOntology::PROPERTY_SOURCE_URI);
+        $class->setPropertyValue($propertySource, $values[tao_actions_form_RemoteList::FIELD_SOURCE_URL]);
+
+        $propertySource = $class->getProperty(RemoteSourcedListOntology::PROPERTY_ITEM_LABEL_PATH);
+        $class->setPropertyValue($propertySource, $values[tao_actions_form_RemoteList::FIELD_ITEM_LABEL_PATH]);
+
+        $propertySource = $class->getProperty(RemoteSourcedListOntology::PROPERTY_ITEM_URI_PATH);
+        $class->setPropertyValue($propertySource, $values[tao_actions_form_RemoteList::FIELD_ITEM_URI_PATH]);
+
+        if ($this->isListsDependencyEnabled()) {
+            $propertySource = $class->getProperty(RemoteSourcedListOntology::PROPERTY_DEPENDENCY_ITEM_URI_PATH);
+            $class->setPropertyValue(
+                $propertySource,
+                $values[tao_actions_form_RemoteList::FIELD_DEPENDENCY_ITEM_URI_PATH]
+            );
+        }
+
+        return $class;
     }
 
-    private function getListCreator(): ListCreator
+    /**
+     * @throws core_kernel_persistence_Exception
+     */
+    private function getListData(bool $showRemoteLists = false): array
     {
-        return $this->getPsrContainer()->get(ListCreator::class);
+        $listService = $this->getListService();
```

Comment:
I would probably create another service like `ListSearcher::search(Query $query): ListPagination` instead of add this logic here.

I would also remove the constants like `REMOTE_LIST_PREVIEW_LIMIT` from the controller.